narwhal/types/src/consensus.rs

```diff
@@ -14,6 +16,68 @@ use tokio::sync::mpsc;
 /// A global sequence number assigned to every certificate.
 pub type SequenceNumber = u64;
 
+/// The output format of the consensus.
+#[derive(Serialize, Deserialize, Clone, Debug, Default, PartialEq)]
+pub struct ConsensusOutput {
+    /// The sequenced certificate.
+    pub certificate: Certificate,
+    /// The (global) index associated with this certificate.
+    pub consensus_index: SequenceNumber,
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, Default)]
+pub struct CommittedSubDag {
+    /// The sequence of committed certificates.
+    pub certificates: Vec<ConsensusOutput>,
+    /// The leader certificate responsible of committing this sub-dag.
+    pub leader: Certificate,
```

Comment:
I am thinking whether we actually need this stored here or we can infer from the `certificates` anyways. I mean, we expect the `certificates` to be ordered and the leader should always be the last "committed"  - so we expect it to be the last certificate of the list. So practically since the DAG is linearised I am thinking whether we would be equally ok of just storing the `certificates` and on the fly get the leader when needed. Then the struct could be reduced to just:

```
struct CommittedSubDag(Vec<ConsensusOutput>)
```

and could have a method like:

```
    pub fn leader(&self) -> Option<ConsensusOutput> {
        self.0.last().cloned()
    }
```

Reply:
Are we sure the leader will be the last committed? I see that we have (scary) modular code to traverse the dag now. I was afraid that we may change it in the future and break safety. Shouldn't we instead try to remove the list of committed certificates from storage?

Reply:
> > I also have a question about wrapping/unwrapping subdag into shell - why did you chose to do it this way? Would it be easier/less error prone if we just had one CF for entire subdag?
> 
> @andll could you please elaborate more on this idea? Wanna make sure I get it right before I comment

Oh I missed this comment somehow. I don't like the sub-dag shell, but it seemed wrong to persist the entire certificate's data. So we only persist digests of the certificates and retrieve the corresponding data upon crash-recovery.